### PR TITLE
Support remote commands in the kernel picker

### LINKED ISSUE #10191: Suggestion: Query clipboard permissions when opening an existing server instead of always prompting
Testing https://github.com/microsoft/vscode-jupyter/issues/10170

I find the setting up Jupyter experience kind of complicated. Have you considered doing a soft permissions query instead of a hard text read when connecting to avoid this prompt? Additionally, it would be better from a user's perspective imo to ask for the clipboard on ctrl/cmd+v, not when clicking some button.

Screen Shot 2022-05-31 at 10 12 11 am

```
navigator.permissions.query({ name: 'clipboard-read' }).then(console.log)
```

### LINKED ISSUE #10363: Choosing/changing remote local servers is buggy and confusing.
Prep
- Start a jupyter server (ala run ```jupyter notebook```)
- Start VS Code Insiders (I used build from 6/7/2022 and corresponding Jupyter extension)
- Create a new notebook
- Select a local kernel and run a simple cell

Bug 1
- Run ```Jupyter: Specify Jupyter server for connections```
- Choose the server you created in the prep.
- Press enter

I'm still connected to my local kernel. I expected to have to choose a new kernel.

Bug 2
- Run ```Notebook: Select notebook kernel```

The suggested kernel will be the local kernel I'm already connected to. I expect to either have no suggestion, or a suggestion from the server I just connected to.
image

Bug 3
- Choose one of the server's kernels, run a cell. All seems well.
- Run ```Notebook: Select notebook kernel``` You'll still see the local kernel suggested. Probably a dupe of Bug 2, but still.
- Select the local kernel and run a cell.

You'll run against the local kernel, yet the status bar says you're connected to a Remote Jupyter server. Highly confusing.

Bug 4
- Choose one of the server kernels again.
- Ctrl+C the Jupyter server started in Prep.
- Run a cell in VS Code (which is still "connected" to the dead server)

Cell "executes" indefinitely and I can't interrupt it.
image

I'm forced to reload.

### LINKED ISSUE #10435: Rework remote support to account for more scenarios
Remote support is confusing:
- [ ] https://github.com/microsoft/vscode-jupyter/issues/10363
- [x] https://github.com/microsoft/vscode-jupyter/issues/10191

Using this issue to describe the use cases.

## Use Cases

1. (web) User opens a notebook and wants to run a cell so they can see the output of a cell
2. (web) User is in a notebook and wants to run a cell in the browser so they don't have to pay for/setup a remote compute
3. (web) User is in a notebook and wants to run a cell on their local machine so they don't have to pay for/setup a remote compute
4. (web) User is in a notebook and wants to run a cell on a different remote server because this one is more expensive
5. (desktop) User opens a notebook and wants to run a cell without entering anything so they can see the output
6. (desktop) User opens a notebook and wants to run it on their own compute so they can use the GPU/CPU on that machine
7. (desktop) User wants to specify their own compute for all notebooks so they don't have to keep picking a server
8. (desktop) User wants to run some cells on local and some cells on remote (need to check telemetry for this)
9. (codespaces) User wants to run a cell and not have to specify anything except clicking the run button.

### PR DESCRIPTION
**Pull request by rchiodo**
Fixes #10435 
Fixes #10363 
Fixes #10191

Refactor remote support such that:
- There's a top level command in the kernel picker for switching to remote
- You can only be in 'remote' or 'local' mode
- The remote picker lets you type the URI right into the first stage
- If you're in remote mode already, you can
  - Enter another URI
  - Go back to local mode
- If you're in local mode, you can enter remote mode by entering a URI
- The 'Jupyter specify server connection' command also is the same URI picker too. No more 'None' and no more 'Existing' step.

Telemetry for these new modes is already handled. Commands send telemetry on their own and remote/local mode is already sent in telemetry.

-   [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [x] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [x] Appropriate comments and documentation strings in the code.
-   [x] Has sufficient logging.
-   [x] Has telemetry for feature-requests.
-   [x] Unit tests & system/integration tests are added/updated.
-   [ ] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [ ] [`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).

### CODE REVIEW

**Comment by rchiodo** [comment]
This can be merged after endgame.

File: src/kernels/jupyter/launcher/serverConnectionType.ts
Comment (rchiodo) [inline src/kernels/jupyter/launcher/serverConnectionType.ts:11]: I deleted this class and moved its functionality to the JupyterServerUriStorage class.

The reason for this is that this class was essentially exposing the internal state of the JupyterServerUriStorage class and they had to be in sync. And they weren't always.

To make this impossible, I just removed this class altogether and added a new interface to the JupyterServerUriStorage class. Most of the changes for this is simply using an interface instead of a class.

Good use case for why we should expose interfaces to the DI container and not classes.

File: src/kernels/jupyter/launcher/serverUriStorage.ts
Comment (rchiodo) [inline src/kernels/jupyter/launcher/serverUriStorage.ts:51]: THis is the code that was originally in the serverConnectionType.ts file.

File: src/kernels/jupyter/serverSelector.ts
Comment (rchiodo) [inline src/kernels/jupyter/serverSelector.ts:149]: Since URI entering happens on the quick pick, we can't use input box validation. So I added this to allow a modal dialog to appear.

File: src/notebooks/controllers/controllerLoader.ts
Comment (rchiodo) [inline src/notebooks/controllers/controllerLoader.ts:186]: All filtering is done in the controllerRegistration.

File: src/platform/common/utils/multiStepInput.ts
Comment (rchiodo) [inline src/platform/common/utils/multiStepInput.ts:176]: I think I need to change this to support markdown. For remote there's a URL the user can click on and it doesn't seem to be clickable at the moment. Maybe dialogs don't support this.

File: src/standalone/activation/activationManager.ts
Comment (rchiodo) [inline src/standalone/activation/activationManager.ts:35]: I had a bug in my activate and it made the extension not start. Figured this was better.

File: src/test/datascience/jupyter/serverSelector.unit.test.ts
Comment (rchiodo) [inline src/test/datascience/jupyter/serverSelector.unit.test.ts:318]: Not supporting the clipboard anymore. Fixes the problem with web where the user may not have clipboard permissions. 

New workflow doesn't really need clipboard as the user can also filter in the same entry so defaulting to the url in the clipboard doesn't really make sense.

File: package.nls.json
Comment (DonJayamanne) [inline package.nls.json:959]: @IanMatthewHuff looks like this will need to be reviewed closly as you're also working in the similar area (installing python when its missing). I.e. you might want to review the workflow & the changes.
Feels like this stuff related to no python might require its own folder/set of classes & its been worked on for months if not years (i.e. workflow keeps changing, hence better to keep it isolated & easier to manage)

Agai, I haven't reviewed anything yet, but just a thought about this (prompting to install python - getting started) part of the exetnsion

File: src/kernels/jupyter/launcher/serverUriStorage.ts
Comment (DonJayamanne) [inline src/kernels/jupyter/launcher/serverUriStorage.ts:35]: We could have issues due to race conditions.
Basically `getUri` is async, hence `currentUri` could be empty when it shouldn't be and could give the wrong impression when its used.

Comment (DonJayamanne) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:35]: Isn't this a duplicate of  the `getUri` method, without the async bit.

File: src/kernels/jupyter/launcher/serverUriStorage.ts
Comment (DonJayamanne) [inline src/kernels/jupyter/launcher/serverUriStorage.ts:41]: I think we can remove `currentUri` and make a sync version of `isLocalLaunch` available by:
* Returning `true` if we're on web
* Retruning `true` if we have a value in the global memento 
     * I.e. when ever we get/update the encrypted store `await this.encryptedStorage.store(Settings.JupyterServerRemoteLaunchService, key, uri);`, then we know we have a remote, at the same time we can update a memento with a flag.
     * & then use that flag to determine if we have a remote or not & thats a synchronous request, hence we don't need promises & we can still have a non-async version of `isLocalLaunch` without accessing the `this.currentUri`

Comment (DonJayamanne) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:41]: Again, as mentioned earlier, if we access this property too early then we might get th ewrong information, e.g. if `currentUri` hasn't been set we'll get `isLocalLaunch` as true even in web cases (which is incorrect).

Comment (DonJayamanne) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:41]: This race condition was resolved by the `ServerConnectionType` class.

File: src/kernels/jupyter/launcher/serverUriStorage.ts
Comment (rchiodo) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:35]: Yes. The sync version. getUri just uses this value once it's set anyway.

File: src/kernels/jupyter/launcher/serverUriStorage.ts
Comment (DonJayamanne) [inline src/kernels/jupyter/launcher/serverUriStorage.ts:58]: I don't think we can assume this will happen very quickly, its possible the encrypted storage is very slow (or could present UI dialog boxes to the user and the like) & the rest of the code will run to completion, leaving the properties like `currentUri` & `isLocalLaunch` in a bad (incorect state).

E.g. in the past when ever we accessed encrypted storage, we'd get a dialog prompting the user to allow accessing secret store, this could happen in the future as well, (.e. this promise could get resolved seconds or minutes later)

File: src/kernels/jupyter/launcher/serverUriStorage.ts
Comment (rchiodo) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:41]: And no it wasn't solved by the ServerConnectionType class. It was merely hidden. ServerConnectionType would say not local while the URI wasn't set yet. So if you tried to get the URI, you'd be off of the actual state.

It's why I removed the ServerConnectionType.

Comment (rchiodo) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:41]: In fact I think we can get rid of the async version. We don't need it. It could be internal only.

Comment (DonJayamanne) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:41]: > say not local while the URI wasn't set yet. So if you tried to get the URI, you'd be off of the actual state.

We can solve this easily by updating the mememto & a private property.
Hence if the private property has been set, then use that else use memento.
When the class is instantiated the first time (ie. loading VS Code), then property will be `undefined` and we fall back to mememto.

& later on everytime we use the property as that will have the latest state.

File: src/kernels/jupyter/launcher/serverUriStorage.ts
Comment (DonJayamanne) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:58]: along with a private property (so that we have sync & latest access to information - even while we're still busy updating the memeto)

Comment (rchiodo) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:58]: Actually isLocalLaunch should use the _local flag. That was a copy of the logic in the ServerConnectionType. I'll update.

File: src/kernels/jupyter/launcher/serverUriStorage.ts
Comment (rchiodo) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:41]: The whole problem was this state that is mixed between two classes. That seemed broken to me. The true state was in the SErverUriStorage, not the ServerConnectionType.

File: src/kernels/jupyter/launcher/serverUriStorage.ts
Comment (DonJayamanne) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:35]: I think we can remove this, we don't need a suync version, we can make this async &
just make other parts of the code async when accessing this.
From what I could tell we use this in adding controllers, and we need to do is turn the `IControllerRegistration.add` method into an async method.
From what I could tell that should be totally fine, as almost every method that calls the `add` method is already async.

File: src/notebooks/controllers/controllerRegistration.ts
Comment (DonJayamanne) [inline src/notebooks/controllers/controllerRegistration.ts:187]: I think you can use `serverId` instead of `baseUrl` that would still allow you to keep all of this synchronous

File: src/kernels/jupyter/launcher/serverUriStorage.ts
Comment (DonJayamanne) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:35]: From what I can tell we only need `currentUri` for filtering:
* We could use the `getUri` async method & turn all others into async
* Or just use the `serverId` (we can save server Id in memto as that's safe & hashed, and its part of the connection metadata, & since thats all synchronous, we can use that via `currentServerId` property & then use that for filtering)

Comment (rchiodo) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:35]: Wait what's serverId? Is that the hash of the connection information.

File: src/notebooks/controllers/controllerRegistration.ts
Comment (rchiodo) [reply src/notebooks/controllers/controllerRegistration.ts:187]: serverId is a hash of the url? Not sure about that. It's not cryptographically secure. I mean if that was allowed, we wouldn't need encrypted storage.

File: src/kernels/jupyter/launcher/serverUriStorage.ts
Comment (rchiodo) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:35]: Hmm. ServerId is a hash of the current uri. It's not secure but it's only in memory. If we wrote it to a memento then there's a risk of leaking somebody's token. I guess there's a risk of leaking the token in memory too, but same with encrypted storage. 

I don't think we can solve this with server id, because we can't write it to the memento.

Comment (rchiodo) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:35]: currentUri being sync is only a problem if controllerRegistration.add is called before it's resolved. 

However if it's resolved and the onDidChangeUri event were to fire, it doesn't matter. It's the same as if it was originally set to remote.

Oh but that still means isLocalLaunch wouldn't match. Hmm. 

So then we'd have to force async.

That ripples through the controllerRegistration add, onDidChangeUri, and on didChangeKernelFilter

- add being async seems possible.
- onDidChangeUri would should work if it actually gets passed the URI involved. 
- onDidChangeKerneFilter could be made async, doesn't really matter

Okay I think I can get rid of the synchronous of the currentUri.

File: src/notebooks/controllers/controllerRegistration.ts
Comment (rchiodo) [reply src/notebooks/controllers/controllerRegistration.ts:187]: Going to try doing this with async instead.

File: src/kernels/jupyter/launcher/serverUriStorage.ts
Comment (DonJayamanne) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:35]: > I don't think we can solve this with server id, because we can't write it to the memento.

Given that we can store it in memento, i think we can solve it

Comment (DonJayamanne) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:35]: > Oh but that still means isLocalLaunch wouldn't match. Hmm.

My suggestion is:
* Create a property to keep track of isloca or remote (initially undefiend)
* When loading the first time, this property is empty and we read the value from memento to get the server id or the like (then we know whether its local or not)
* Subsequently when we update memento, we can update this property as well
* Then next time we can check if this property is not undefined & use the value in that property.
Thats how the serverConnectionType class worked, i.e. it solved this problem by having synchronous access via the `_isLocalLaunch` private property.

File: src/notebooks/controllers/controllerRegistration.ts
Comment (DonJayamanne) [reply src/notebooks/controllers/controllerRegistration.ts:187]: > The difficult part is an update after onDidChangeUri. Maybe if I pass the URI there and during add, I can skip the sync 'currentUri'

See comments above, the previous class `ServerConnectionType._isLocalLaunch`

File: src/kernels/jupyter/launcher/serverUriStorage.ts
Comment (rchiodo) [reply src/kernels/jupyter/launcher/serverUriStorage.ts:35]: hashjs is not cryptographically secure. We'd have to use something else. The security review team specifies this when we said we don't use it to store anything but just to hash stuff that isn't a password. We start using it to store passwords (which is what the token in the url is) we need to use something other than hashjs.

File: package.nls.json
Comment (IanMatthewHuff) [reply package.nls.json:959]: @DonJayamanne I think Rich was just adding on the Command localization that I missed in my addition. The only real conflict to consider here is just the fact that we won't have only one command anymore in the new user scenario. Means that we don't get that nice "just hit the run button" launch, but we can tweak that as we move forward. Maybe prioritization for commands in the kernel picker somehow.

File: src/notebooks/controllers/commands/serverConnectionControllerCommands.ts
Comment (IanMatthewHuff) [inline src/notebooks/controllers/commands/serverConnectionControllerCommands.ts:240]: Is the 'z' for alphabetizing? Doesn't seem like that would work for non-en.

File: src/platform/common/utils/multiStepInput.ts
Comment (IanMatthewHuff) [reply src/platform/common/utils/multiStepInput.ts:176]: I think they don't sadly. Our "Install Python" popup tries to link to Python.org, but it's not clickable. Pretty sure when we talked about it in v-team that's just how it is.

**Comment by IanMatthewHuff** [comment]
Seems good to me. This is going to be a big improvement. Honestly the hardest part of this PR was that I find the MultiStep input control flow really wacky and hard to read. But that's not anything on you or this PR, I think I just need to play with that myself a bit versus looking at the PR.

File: src/notebooks/controllers/commands/serverConnectionControllerCommands.ts
Comment (rchiodo) [reply src/notebooks/controllers/commands/serverConnectionControllerCommands.ts:240]: I copied this from the core, but it does seem wrong. 'z' only works in ascii.

**Comment by rchiodo** [comment]
> Seems good to me. This is going to be a big improvement. Honestly the hardest part of this PR was that I find the MultiStep input control flow really wacky and hard to read. But that's not anything on you or this PR, I think I just need to play with that myself a bit versus looking at the PR.

I agree about the multistep input. It's really a state machine. I think it would be better with an iterator type pattern instead of a promise type pattern. Took me a long time to get the promises to work out correctly.

File: src/notebooks/controllers/commands/serverConnectionControllerCommands.ts
Comment (rchiodo) [reply src/notebooks/controllers/commands/serverConnectionControllerCommands.ts:240]: I'll enter an issue to relook at this in both the core and here.

File: src/kernels/jupyter/serverSelector.ts
Comment (rchiodo) [reply src/kernels/jupyter/serverSelector.ts:160]: Yeah there's only two buttons for the server selector. Back and the 'delete' this URI.

**Review by DonJayamanne — APPROVED** [review]

**Review by IanMatthewHuff — APPROVED** [review]

File: src/kernels/jupyter/serverSelector.ts
Comment (sadasant) [inline src/kernels/jupyter/serverSelector.ts:59]: Are these not needed anymore?

Comment (rchiodo) [reply src/kernels/jupyter/serverSelector.ts:59]: No they shouldn't be. Maybe. I think we might need the 'None' label.

File: src/notebooks/controllers/commands/serverConnectionControllerCommands.ts
Comment (sadasant) [inline src/notebooks/controllers/commands/serverConnectionControllerCommands.ts:43]: Ohh fancy! Would it make sense to have a common place for functional utilities like this?

File: src/test/datascience/debugger.vscode.test.ts
Comment (sadasant) [inline src/test/datascience/debugger.vscode.test.ts:121]: Is this necessary?

File: src/test/datascience/interactiveDebugging.vscode.common.ts
Comment (sadasant) [inline src/test/datascience/interactiveDebugging.vscode.common.ts:101]: Should we have an issue to re-enable this test?

File: src/test/datascience/interactiveWindow.vscode.test.ts
Comment (sadasant) [inline src/test/datascience/interactiveWindow.vscode.test.ts:103]: Should we have an issue to re-enable this test?

File: src/test/datascience/interactiveDebugging.vscode.common.ts
Comment (sadasant) [reply src/test/datascience/interactiveDebugging.vscode.common.ts:101]: I see the issue! Disregard.

File: src/test/datascience/debugger.vscode.test.ts
Comment (rchiodo) [reply src/test/datascience/debugger.vscode.test.ts:121]: Test was failing attempting to see if files were cleaned up. Stopped failing after I did this, so yeah probably.

There's likely another way to accomplish this but I wasn't working on this code so I just wanted the test to pass

File: src/notebooks/controllers/commands/serverConnectionControllerCommands.ts
Comment (rchiodo) [reply src/notebooks/controllers/commands/serverConnectionControllerCommands.ts:43]: Maybe. Not sure where that would be. It's only used in this class at the moment.

File: src/test/datascience/debugger.vscode.test.ts
Comment (sadasant) [reply src/test/datascience/debugger.vscode.test.ts:121]: Ok that makes sense! Thanks for the context

**Pull request merged**
